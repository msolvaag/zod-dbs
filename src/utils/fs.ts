import { existsSync, readdirSync, unlinkSync } from 'fs';
import { mkdir } from 'fs/promises';
import path from 'path';

import { ZodPgTableType } from '../types.js';
import { logDebug } from './debug.js';

/**
 * Deletes all .ts files in the output schemas folder.
 */
export function clearTablesDirectory(outputPath: string) {
  // Check if the directory exists first
  if (!existsSync(outputPath)) {
    logDebug(`Directory ${outputPath} does not exist, nothing to clear`);
    return;
  }

  const files = readdirSync(outputPath, {
    recursive: true,
    withFileTypes: true,
  });

  for (const file of files) {
    if (file.isDirectory()) continue;
    if (!file.name.endsWith('.ts')) continue;

    const filePath = path.join(file.path, file.name);
    unlinkSync(filePath);
  }

  if (files.length > 0) {
    logDebug(`Deleted all .ts files in ${outputPath}`);
  }
}

export async function ensureFolder(folderPath: string) {
  const exists = existsSync(folderPath);

  if (!exists) {
    await mkdir(folderPath, { recursive: true });
  }
}

export const getOutputFolder = (type: ZodPgTableType): string => {
  switch (type) {
    case 'table':
      return 'tables';
    case 'materialized_view':
      return 'materialized_views';
    case 'view':
      return 'views';
    case 'foreign_table':
      return 'foreign_tables';
    default:
      return 'others';
  }
};
