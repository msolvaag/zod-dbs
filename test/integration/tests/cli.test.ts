import { execSync } from 'child_process';
import fs from 'fs';
import path from 'path';

import {
  getClientConnectionString,
  getOutputFiles,
  outputDir,
} from '../testDbUtils';

const cliPath = path.resolve(__dirname, '../../../index.js');

it('CLI generates correct zod schemas', () => {
  const connectionString = getClientConnectionString();

  execSync(
    `node ${cliPath} --connection-string "${connectionString}" --output "${outputDir}" --json-schema-import-location "../../json"`,
    { stdio: 'inherit' }
  );

  const outputFiles = getOutputFiles();

  for (const file of outputFiles) {
    const content = fs.readFileSync(file, 'utf8');
    expect(content).toMatchSnapshot(path.relative(outputDir, file));
  }
});
