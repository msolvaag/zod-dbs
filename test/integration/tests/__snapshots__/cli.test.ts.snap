// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CLI generates correct zod schemas: constants.ts 1`] = `
"// This file is auto-generated by zod-pg
// Do not edit this file directly.

export const TABLE_POSTS = 'posts';
export const TABLE_USERS = 'users';
"
`;

exports[`CLI generates correct zod schemas: tables/index.ts 1`] = `
"// This file is auto-generated by zod-pg
// Do not edit this file directly.

export * from './posts';
export * from './users';
"
`;

exports[`CLI generates correct zod schemas: tables/posts.ts 1`] = `
"// This file is auto-generated by zod-pg
// Do not edit this file directly.

import { z } from 'zod';


export const PostsTableInsertSchema = z.object({
  user_id: z.number().int().nullable().optional(),
  title: z.string(),
  content: z.string().nullable().optional(),
  published: z.boolean().nullable().optional(),
});

export const PostsTableSchema = z.object({
  id: z.number().int(),
  user_id: z.number().int().nullable().optional().transform(val => (val === null ? undefined : val)),
  title: z.string(),
  content: z.string().nullable().optional().transform(val => (val === null ? undefined : val)),
  published: z.boolean().nullable().optional().transform(val => (val === null ? undefined : val)),
});

export const PostsTableUpdateSchema = PostsTableInsertSchema.partial();

export type PostRecord = z.infer<typeof PostsTableSchema>;
export type PostInsertRecord = z.input<typeof PostsTableInsertSchema>;
export type PostUpdateRecord = z.input<typeof PostsTableUpdateSchema>;
"
`;

exports[`CLI generates correct zod schemas: tables/users.ts 1`] = `
"// This file is auto-generated by zod-pg
// Do not edit this file directly.

import { z } from 'zod';
import * as json from '../../json';


export const UsersTableInsertSchema = z.object({
  name: z.string().max(100),
  email: z.string().max(255).nullable().optional(),
  status: z.string().max(20).nullable().optional(),
  profile: json.UserProfileSchema.nullable().optional().transform(val => val ? JSON.stringify(val) : null),
  created_at: z.date({ coerce: true }).nullable().optional(),
});

export const UsersTableSchema = z.object({
  id: z.number().int(),
  name: z.string(),
  email: z.string().nullable().optional().transform(val => (val === null ? undefined : val)),
  status: z.string().nullable().optional().transform(val => (val === null ? undefined : val)),
  profile: json.UserProfileSchema.nullable().optional().transform(val => (val === null ? undefined : val)),
  created_at: z.date({ coerce: true }).nullable().optional().transform(val => (val === null ? undefined : val)),
});

export const UsersTableUpdateSchema = UsersTableInsertSchema.partial();

export type UserRecord = z.infer<typeof UsersTableSchema>;
export type UserInsertRecord = z.input<typeof UsersTableInsertSchema>;
export type UserUpdateRecord = z.input<typeof UsersTableUpdateSchema>;
"
`;

exports[`CLI generates correct zod schemas: types.ts 1`] = `
"// This file is auto-generated by zod-pg
// Do not edit this file directly.

export type Table = 
  | 'posts'
  | 'users';
"
`;
